### PR TITLE
Import ClangSA.jl source tree

### PR DESCRIPTION
**Pull request by Keno**
Now that static analysis passes on src/*.c without any false positives,
it's time to integrate ClangSA into our regular CI process. Since it's
a bit specific to the exact version of julia, and for convenience,
we import the source tree into julia proper. `make -C src analyzegc`
runs the analysis. ClangSA also comes with its own test suite, which
is imported, but not hooked up (To be done as part of #28330).

### CODE REVIEW

File: src/clangsa/Makefile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+SRCDIR := $(abspath $(dir $(lastword $(MAKEFILE_LIST))))
+JULIAHOME := $(abspath $(SRCDIR)/../..)
+BUILDDIR := .
+include $(JULIAHOME)/deps/Versions.make
+include $(JULIAHOME)/Make.inc
+include $(JULIAHOME)/deps/llvm-ver.make
```
Comment (vchuravy) [inline src/clangsa/Makefile]: Same here and I think you only need Make.inc

**Comment by Keno** [comment]
Talking with @vchuravy, it probably makes sense to try to run the clangsa as a separate travis worker. We use the regular travis worker to test our LLVM build, plus the static analysis takes a while (~5-10mins per platform - though we only do x86_64 for now), but most importantly, it's easy to get failures here, so we probably want it to fail as fast as possible if we can.

**Comment by ararslan** [comment]
> it probably makes sense to try to run the clangsa as a separate travis worker.

Our use of CircleCI is redundant with Travis. Perhaps we can repurpose a Circle build for that?

**Comment by Keno** [comment]
> Our use of CircleCI is redundant with Travis. Perhaps we can repurpose a Circle build for that?

Maybe, but it has been nice to have Circle when travis inevitably has outages. Also it turns out that if we don't do any apt installs, we get to starting the clangsa part of this in about 1 min of travis time, which is significantly better than I thought.

**Comment by c42f** [comment]
I've been trying to use this to understand some CI failures in https://github.com/JuliaLang/julia/pull/28878 by merging this branch with that.

After a local build (but without running the install step), `make -C src analyzegc` fails because `clang` isn't available in `$JULIA_SRC_ROOT/usr/tools/clang`. Is this expected? I noticed that `usr/tools/clang++` and `usr/tools/clang-cl`, `usr/tools/clang-cpp` are installed as symlinks which point to the nonexistent `usr/tools/clang`.

By the way I'm looking forward to having this merged!

**Comment by c42f** [comment]
Got it working for now with some symlinks, and it told me where a `GC_POP` was missing :100:

What's left to do to get this merged? If it's CI integration which is holding this up I'd be in favor of merging a version without the CI changes and doing those as a next step.

**Comment by vchuravy** [comment]
I think the LLVMBuilder should now be able to be used for this? So this should only need a rebase.

**Comment by Keno** [comment]
I'll rebase this and get it merged without the changes to travis.yml. Afterwards we need to get the analysis clean again and then we can enable it on CI.

**Comment by vtjnash** [comment]
Seems worthy of a NEWS entry too?

**Comment by c42f** [comment]
Having this in is so great :tada:

**Pull request merged**